xarray/tests/test_concat.py

```diff
@@ -86,7 +783,7 @@ def test_concat_merge_variables_present_in_some_datasets(self, data) -> None:
         split_data = [data.isel(dim1=slice(3)), data.isel(dim1=slice(3, None))]
         data0, data1 = deepcopy(split_data)
         data1["foo"] = ("bar", np.random.randn(10))
-        actual = concat([data0, data1], "dim1")
+        actual = concat([data0, data1], "dim1", data_vars="minimal")
```

Comment:
Is this change necessary?

Reply:
That's kind of corner case here which reveals a bug using `ensure_common_dims`.

`ensure_common_dims` expects `variables` having the same length as `concat_dim_lenghts`. But that's not true for `foo` wich is only available in the second dataset. Result is that `foo`'s `dim1` gets the wrong length (3 instead of 5 in this case).

We could solve this by providing the correct `concat_dim_lenghts` as argument to `ensure_common_dims`.

Reply:
I've added a proposal how to fix this.